### PR TITLE
doc: add design proposal of the out-of-tree node IPAM controller

### PR DESCRIPTION
**Pull request by nilo19**
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespaces from that line:
>
> /kind api-change
> /kind bug
> /kind cleanup
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake
/kind design

**What this PR does / why we need it**:
Add design proposal of the out-of-tree node IPAM controller

**Which issue(s) this PR fixes**:

Fixes #
Ref: https://github.com/kubernetes-sigs/cloud-provider-azure/issues/462

**Special notes for your reviewer**:

**Release note**:
```
NONE
```

### CODE REVIEW

File: site/content/en/development/design-docs/_index.md [hunk 1]
```diff
@@ -6,3 +6,6 @@ weight: 4
 description: >
     Design Docs and KEPs related to this project.
 ---
+
+This is the staging area of the design docs prior to or under development. Once the feature is done, the corresponding 
+design doc would be moved to [Topics](../../topics)
```
Comment (feiskyer) [inline site/content/en/development/design-docs/_index.md:11]: missing `.` at the end.

File: site/content/en/development/design-docs/ipam.md [hunk 2]
```diff
@@ -0,0 +1,38 @@
+---
+title: "Support node IPAM controller"
```
Comment (feiskyer) [inline site/content/en/development/design-docs/ipam.md:2]: Rename title to "Out-of-tree Node IPAM Controller" (to distinguish from in-tree ipam controller)

File: site/content/en/development/design-docs/ipam.md [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+linkTitle: "node IPAM controller"
+type: docs
+weight: 10
+description: >
+Support node IPAM controller in the out-of-tree provider Azure.
+---
+
+## Background
+
+The in-tree node IPAM controller would be moved to the GCE staging repo in the future,
```
Comment (feiskyer) [inline site/content/en/development/design-docs/ipam.md:12]: The in-tree [Node IPAM controller](https://github.com/kubernetes/kubernetes/tree/master/pkg/controller/nodeipam) only supports a fixed node CIDR mask size for all nodes, while in multiple node pool (VMSS) scenarios, different mask sizes are required for different node pools. There is a GCE-specific cloud CIDR allocator for a similar scenario, but that is not exposed in cloud provider API and it is planned to be moved out-of-tree. Hence this docs proposes an out-of-tree node IPAM controller.

File: site/content/en/development/design-docs/ipam.md [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+[PR](https://github.com/kubernetes/kubernetes/pull/97925). Hence, we need to copy the code to the out-of-tree repo and 
+implement the azure-specific logic here. Specifically, we want to allocate the CIDR for each VMSS and VMAS so every
+instance would get a pod CIDR within the range of the corresponding VMSS or VMAS, instead of the cluster CIDR.
+
+## User experience
+
+There are two kinds of CIDR allocator in the node IPAM controller, which are `RangeAllocator` and `CloudAllocator`.
+The `RangeAllocator` is the default one which allocates the pod CIDR for every node in the range of the cluster CIDR.
+The `CloudAllocator` allocates the pod CIDR for every node in the range of the CIDR on the corresponding VMSS or VMAS.
+The CIDR on the VMSS or VMAS is set by a specific tag `{"VMSet-CIDR": "10.244.1.0/24"}` and every node would get a
```
Comment (feiskyer) [inline site/content/en/development/design-docs/ipam.md:22]: Let's only add the size in VMSS/VMAS tag, e.g. `"kubernetesNodeCIDRMaskSize": "24"`

File: site/content/en/development/design-docs/ipam.md [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+
+There are two kinds of CIDR allocator in the node IPAM controller, which are `RangeAllocator` and `CloudAllocator`.
+The `RangeAllocator` is the default one which allocates the pod CIDR for every node in the range of the cluster CIDR.
+The `CloudAllocator` allocates the pod CIDR for every node in the range of the CIDR on the corresponding VMSS or VMAS.
+The CIDR on the VMSS or VMAS is set by a specific tag `{"VMSet-CIDR": "10.244.1.0/24"}` and every node would get a 
+pod CIDR within the range. Note that the range in the VMSS or VMAS must be within the cluster CIDR, or an error would be 
+thrown. The mask of each pod CIDR could be modified by setting `--node-cidr-mask-size`. For example, 
+if the `--node-cidr-mask-size` is 24 the first node would be allocated with a pod CIDR of "10.244.1.0/24", and the second
+would be allocated "10.244.2.0/24", etc.
+
```
Comment (feiskyer) [inline site/content/en/development/design-docs/ipam.md:27]: Add a comment that `--node-cidr-mask-size` would be the default mask size when the above tag doesn't exist.

File: site/content/en/development/design-docs/ipam.md [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+| name | type | default | description |
+| ----- | -----| ----- | ----- |
+| allocate-node-cidrs | bool | true | Should CIDRs for Pods be allocated and set on the cloud provider. |
+| cluster-cidr | string | "10.244.0.0/16" | CIDR Range for Pods in cluster. Requires --allocate-node-cidrs to be true |
+| service-cluster-ip-range | string | "" | CIDR Range for Services in cluster. Requires --allocate-node-cidrs to be true |
+| node-cidr-mask-size | int | 24 | Mask size for node cidr in cluster. Default is 24 for IPv4 and 64 for IPv6. |
+| node-cidr-mask-size-ipv4 | int | 24 | Mask size for IPv4 node cidr in dual-stack cluster. Default is 24. |
+| node-cidr-mask-size-ipv6 | int | 64 | Mask size for IPv6 node cidr in dual-stack cluster. Default is 64. |
+| cidr-allocator-type | string | "RangeAllocator" | The CIDR allocator type. "RangeAllocator" or "CloudAllocator" |
```
Comment (feiskyer) [inline site/content/en/development/design-docs/ipam.md:38]: open question: what would happen if the tag value gets changed after some time? 

In this case, I think we shouldn't change the nodeCIDR for existing nodes, but new nodes' CIDR should be allocated with a new mask size.

**Review by feiskyer — APPROVED** [review]

File: site/content/en/development/design-docs/ipam.md (hunk 3 above)
Comment (nilo19) [reply site/content/en/development/design-docs/ipam.md:12]: actually the custom mask size is support in the in-tree node ipam controller by a kcm flag `node-cidr-mask-size`

**Pull request merged**